pkg/reconciler/internal/paramvalidation/paramvalidation.go

```diff
@@ -0,0 +1,146 @@
+				for key := range spec.Default.ObjectVal {
+					providedKeys[spec.Name] = append(providedKeys[spec.Name], key)
+				}
+			}
+		}
+	}
+
+	// collect provided keys from run level value
+	for _, p := range params {
+		if p.Value.Type == v1beta1.ParamTypeObject {
+			for key := range p.Value.ObjectVal {
+				providedKeys[p.Name] = append(providedKeys[p.Name], key)
+			}
+		}
+	}
+
+	return findMissingKeys(neededKeys, providedKeys)
+}
+
+// findMissingKeys checks if objects have missing keys in its providers (taskrun value and default)
+func findMissingKeys(neededKeys, providedKeys map[string][]string) map[string][]string {
+	missings := map[string][]string{}
+	for p, keys := range providedKeys {
+		if _, ok := neededKeys[p]; !ok {
+			// Ignore any missing objects - this happens when object param is provided with default
+			continue
+		}
+		missedKeys := list.DiffLeft(neededKeys[p], keys)
+		if len(missedKeys) != 0 {
+			missings[p] = missedKeys
+		}
+	}
+
+	return missings
+}
+
+// ExtractParamArrayLengths extract and return the lengths of all array params
+// Example of returned value: {"$(params.array-params)": 2, "$(params[\"array-params\"])": 2, "$(params['array-params'])": 2}
+func ExtractParamArrayLengths(defaults []v1beta1.ParamSpec, params []v1beta1.Param) map[string]int {
+	// Collect all array params
+	arrayParamsLengths := make(map[string]int)
+
+	patterns := []string{
+		"$(params.%s)",
+		"$(params[%q])",
+		"$(params['%s'])",
+	}
+
+	// Collect array params lengths from defaults
+	for _, p := range defaults {
+		if p.Default != nil {
+			if p.Default.Type == v1beta1.ParamTypeArray {
+				for _, pattern := range patterns {
+					for i := 0; i < len(p.Default.ArrayVal); i++ {
+						arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
+					}
+				}
+			}
+		}
+	}
+
+	// Collect array params lengths from params
+	for _, p := range params {
+		if p.Value.Type == v1beta1.ParamTypeArray {
+			for _, pattern := range patterns {
+				for i := 0; i < len(p.Value.ArrayVal); i++ {
+					arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
+				}
+			}
+		}
+	}
+	return arrayParamsLengths
+}
+
+// ValidateOutofBoundArrayParams validates if the array indexing params are out of bound
+func ValidateOutofBoundArrayParams(arrayIndexingParams []string, arrayParamsLengths map[string]int) error {
+	outofBoundParams := sets.String{}
+	for _, val := range arrayIndexingParams {
+		indexString := substitution.ExtractIndexString(val)
+		idx, _ := substitution.ExtractIndex(indexString)
+		v := substitution.TrimArrayIndex(val)
+		if paramLength, ok := arrayParamsLengths[v]; ok {
+			if idx >= paramLength {
+				outofBoundParams.Insert(val)
+			}
+		}
+	}
+	if outofBoundParams.Len() > 0 {
+		return fmt.Errorf("non-existent param references:%v", outofBoundParams.List())
+	}
+	return nil
+}
+
+// ExtractArrayIndexingParamRefs extract the array indexing references
```

Comment:
```suggestion
// ExtractArrayIndexingParamRefs takes a string of the form `foo-$(params.array-param[1])-bar` and extracts the portions of the string that reference an element in an array param.
// For example, for the string ``foo-$(params.array-param[1])-bar-$(params.other-array-param[2])-$(params.string-param)`,
// it would return ["$(params.array-param[1])", "$(params.other-array-param[2])"].
```